### ISSUE TITLE
[clang][analyzer] core.NonNullParamChecker false positive after repeated array-element loads

### ISSUE THREAD

**Issue by bozicrHT** [llvm/llvm-project#223400]
**Description:**
Analyzer reports that the second argument of `memcmp()` may be null, even though the immediately enclosing condition establishes that both arguments have the same nullness and that the first argument is non-null.

**Reproducer:**
```c
#include <string.h>

void compare(unsigned left_index, unsigned right_index)
{
    const char *left[2] = { 0 }, *right[2] = { 0 };

    if (left_index < 2)
        left[left_index] = "a";
    if (right_index < 2)
        right[right_index] = "b";

    for (unsigned i = 0; i < 2; ++i) {
        if ((!left[i] == !right[i]) && left[i]) {

          clang_analyzer_eval(!left[i] == !right[i]); // UNKNOWN
          clang_analyzer_eval(left[i] != NULL); // UNKNOWN
          clang_analyzer_eval(right[i] != NULL); // FALSE and UNKNOWN
        
          (void)memcmp(left[i], right[i], 1);
        }
    }
}
```
The condition
```c
(!left[i] == !right[i]) && left[i]
```
establishes:
1. `left[i]` and `right[i]` have the same nullness.
2. `left[i]` is non-null.
Therefore, `right[i]` must also be non-null.

**Workaround:**
Loading the elements into local variables before testing them prevents the warning:

```c
for (unsigned i = 0; i < 2; ++i) {
    const char *l = left[i];
    const char *r = right[i];

    if ((!l == !r) && l)
        (void)memcmp(l, r, 1);
}
```

Compiler explorer link can be found [here.](https://compiler-explorer.com/z/zra5qMn8r)

**Comment by NagyDonat** [comment]
This is strange – I would've guessed that the `eagerly-assume` logic (which is enabled by default) splits the execution paths when it sees the `!` operators – and that should have been sufficient to guarantee correct behavior here.

**Update:** the mistake probably happens in the modeling of the `LValueToRValueCast` (i.e. "load the value stored at this address") operation. The lvalues of the expressions `left[i]` and `right[i]` are evaluated correctly as element regions within `left` and `right`, but the Load operation apparently produces `0 (Loc)` and `Unknown` (even before the `if` condition that would trigger `eagerly-assume`).

Instead of `Unknown`, the load should produce some sort of symbolic value representing the accessed element (which stays the same if `left[i]` appears twice in the source code).

See also my [tweaked godbolt example](https://compiler-explorer.com/#z:OYLghAFBqd5QCxAYwPYBMCmBRdBLAF1QCcAaPECAMzwBtMA7AQwFtMQByARg9KtQYEAysib0QXACx8BBAKoBnTAAUAHpwAMvAFYTStJg1AB9U8lJL6yAngGVG6AMKpaAVxYMQAZgCcpBwAyeAyYAHLuAEaYxBIAHKQADqgKhLYMzm4e3n5JKTYCQSHhLFExXPGWmNZpQgRMxAQZ7p6%2BFphW%2BQy19QSFYZHRcRZ1DU1ZrQojvcH9JYPlAJQWqK7EyOwcAKQATF7ByG5YANSbXo6TxMHAAHQIp9ibGgCCj08Abqh46EcHhsDGhjEAE8AF7RYyYN5iCDBAgLU4AIVeHy%2BPwMRgBzFooPBAHd6gwAJJUABKmCYyAQmHQEBR6HhXiRzzpaL%2BmOBYOIxnQ7gStM%2B3wAVAyme8BayMYDsZzubzjMhkHy0AxJj8EPUjsLEcjxb9JVicVyeSwEsZXCq8MAQjTzSkrdSRa8dai0Cb6pgILbLdajvQqARjMEsKpSEcvfbvpdgAgA0HMKp4c9NgB2UVHdM/ASqykawV%2BgibACsCO2RYAIicvBWUwijhoTsmy6HBVGY0WS%2BXK9XU3WG2XtUnnhmjngqEcIPnAwxg5XHEdtomnsPh/n25O46pO6dq9ttkwdqXGa9h6Px63Y9P47P54vlxnz%2B3z1Pg1uqyddxEDwOXrtT1xjxm/DEOO4Y%2BngXZ1oiI7XoetY7Ei2wIngCwNmmd4Znq/xSoasommaFoRjCjpDuh6aYey0rgsafI7AAbKuxZ4OWxFLqREpYQaMrURAdEPoxzHfmx7EUTh1HyoqE6YP67ZMYWZYsUJ5HYVxcoKnyfFIQJR6Dqxd6nlAYBgAxmlyV225HIZGmyfJ760XRvpSQW/FyShNYAaRSmcVRcqgdSRGCWxnkct5Jo8ds9GOTJWloR56IccFRq8mFtFWdF7noUFlGJXhamSdJznyQFsVsspIWmrlqUud%2B6V3vyXwLGwLDIKFxnWaGlVNkcXAKehKb9iRd4RMQ5IANZQQA9ONRwAFKuKqRBHMNPLrEcBBUkcrArAQRyoGO1EKOlfVOrs7RKOlQEgQRYEQRoUHgacc6we%2BCFIa5qY1RhWbbTmwF5hBrXlkVy7Ktm6q/cB5kdUDw4g99YOarQ2z/ZFBXQ59Kpw7mxBI5Dlptqj2m6cu%2BkQIZtBmW%2BlmueF9m0G9MWBXFIkqXhvk0shaN3plolJXTnPA0zpXZepPVCZNRyhKgNgrWtTDbWtmAbV5wHoKgmAKAwYAcNt5pYMQkyGN8svbeT26Q0r3yIxTFbYyAH1c4Lyu4XyiOi4zJVO9x2Msfby51fSjXNS77Wht1/NHQNy5DaNE1TbN82oIt1KuDL62ba4227Uc%2B2HY2x1eA4o5Oo2HBLLQnCFrwngcFopCoJwc4KCsayKzsXg8KQBCaKXSwjSAhYaPonCSFX3d15wvAKCAg9dzXpekHAsAwIgKCoCadDROQlCugkG8xAqTMEMQ5ojXwdAENEU8QBEY8RME9RApwHd38wxBAgA8hE2hVLPHeumwgh34MGxGPLAERXDAEcGIWgU9uC8CwCwP44g56kHwMNaobx1Zj3jFUDOGwO6wnaGPWgeAhoP2cFgMeR88AsCfrwTBxAIjJEwGWTAiCjAkKMN3JYVADDAAUAANTwJgXE78EiMDoTIQQIgxDsCkFI%2BQSg1Bj10FwfQfwQCmHlPoUhU9IBLFQAkTosCAC0JjSpHBMQADSFpYoWJjKRVBGtEbcaBhqoMwOAm42BVAJGIISFU4jqgCF4KgBhlwsB6IgEsSowTPAQAcGMTwajAgzGKKUPQuRUgCCSZk5I2SGB9HSfMNoHQahTFyWo2JnRugNCKQMMowweiVKaXUtJDSJAxObusTpQ8OCV1INXWu9cOBHFULEWiJjaKSHYuOI%2BJ8UIQFwIQEg7524LF4LPLQCxe790HuXDgI9SC0K8JIa42xCw%2BAuVcy51zwqDLHiMye09O7cIXsvCASAtoJAzlvCAO896hFYBscZkzpmzN3NcLg1xa7UhWREvQ/BpGiHEPIpFiiVDqBQao0guJiBMASHQsuFdR4oJGe/DOPzM5jlBVMmZmE5nHwYCNRZzh170AhrsbqmzuFLCpEwPWlBiWHN4LQ2ihZrjJkkBoKVMrpWyuTIWB5ZKJ4WBeVsnupA%2B4Dz6V4UlwzVUap2X07Y%2BrQmGt5aQBhKQ7CSCAA%3D%3D%3D) which also shows that the workaround is only a surface-level fix of the issue: when the code introduces `const char *l = left[i];` the analyzer conjures a stable symbolic value for this variable – but `const char *l2 = left[i];` will get a _different_ conjured value, so the analyzer is not aware that `left[i]` produces the same value consistently (while `left` and `i` are unchanged).

This bug is in the `RegionStore` (the memory model used by the analyzer), which is a very complex beast, but I feel that this particular bug could be resolved by a well-targeted local patch that doesn't disturb the overall architecture. Perhaps @steakhal might add something more because he's more familiar with the `RegionStore`.

If somebody is interested in fixing this bug, then I can probably help with answering a few questions and reviewing the PR that they produce, but I probably won't have time to fix this myself.

**Comment by steakhal** [comment]
Yep, this is a RegionStore limitation. Currently, when binding to an address with symbolic offset, we invalidate the whole array (object) with an Unknown default binding. The reasoning is that we can't track where did the symbolic offset point, thus it could be anywhere.
This is sound, but causes trouble like this.

An alternative solution would be to bind a fresh symbol as a default binding instead - similar to what we do for scalars in the Engine `ExprEngine::VisitDeclStmt`. I'm pretty sure that this would be also sound to do.

The problem is that we don't have everything to conjure symbols from the Store, so those must be smuggled into the context (`StoreManager::Bind`) via parameters and there are about ~19 call sites to `ProgramState::bindLoc` that would need to be changed.

Here is the crude patch used for the investigation - made by claude:
```diff
diff --git a/clang/include/clang/StaticAnalyzer/Core/PathSensitive/Store.h b/clang/include/clang/StaticAnalyzer/Core/PathSensitive/Store.h
index cda21138fe5e..24ea3042a25f 100644
--- a/clang/include/clang/StaticAnalyzer/Core/PathSensitive/Store.h
+++ b/clang/include/clang/StaticAnalyzer/Core/PathSensitive/Store.h
@@ -114,6 +114,12 @@ public:
   ///   by \c val bound to the location given for \c loc.
   virtual BindResult Bind(Store store, Loc loc, SVal val) = 0;
 
+  /// PoC(#223400): supply the context needed to conjure a fresh symbol when a
+  /// bind at a symbolic offset has to forget the old contents of the enclosing
+  /// concrete region.
+  virtual void setBindConjureContext(std::optional<ConstCFGElementRef> Elem,
+                                     const StackFrame *SF, unsigned Count) {}
+
   /// Return a store with the specified value bound to all sub-regions of the
   /// region. The region must not have previous bindings. If you need to
   /// invalidate existing bindings, consider invalidateRegions().
diff --git a/clang/lib/StaticAnalyzer/Core/ExprEngine.cpp b/clang/lib/StaticAnalyzer/Core/ExprEngine.cpp
index e482cb7ef7a5..361c04e74121 100644
--- a/clang/lib/StaticAnalyzer/Core/ExprEngine.cpp
+++ b/clang/lib/StaticAnalyzer/Core/ExprEngine.cpp
@@ -3791,10 +3791,16 @@ void ExprEngine::evalBind(ExplodedNodeSet &Dst, const Stmt *StoreE,
     State = processPointerEscapedOnBind(State, Location, Val, SF);
 
     if (auto AsLoc = Location.getAs<Loc>()) {
+      // PoC(#223400): give RegionStore what it needs to conjure a fresh symbol
+      // if this bind lands at a symbolic offset.
+      getStateManager().getStoreManager().setBindConjureContext(
+          getCFGElementRef(), SF, getNumVisitedCurrent());
       // When binding the value, pass on the hint that this is a
       // initialization. For initializations, we do not need to inform clients
       // of region changes.
       State = State->bindLoc(*AsLoc, Val, SF, /*notifyChanges=*/!AtDeclInit);
+      getStateManager().getStoreManager().setBindConjureContext(std::nullopt,
+                                                                nullptr, 0);
     }
 
     PostStore PS(StoreE, SF, Location.getAsRegion(), /*tag=*/nullptr);
diff --git a/clang/lib/StaticAnalyzer/Core/RegionStore.cpp b/clang/lib/StaticAnalyzer/Core/RegionStore.cpp
index 6994917470ce..853ce5aed608 100644
--- a/clang/lib/StaticAnalyzer/Core/RegionStore.cpp
+++ b/clang/lib/StaticAnalyzer/Core/RegionStore.cpp
@@ -519,6 +519,21 @@ private:
     return StateMgr.getOwningEngine().getAnalysisManager().options;
   }
 
+  // PoC(#223400): context for conjuring a fresh default symbol on a
+  // symbolic-offset bind.
+  std::optional<ConstCFGElementRef> BindElem;
+  const StackFrame *BindSF = nullptr;
+  unsigned BindCount = 0;
+
+public:
+  void setBindConjureContext(std::optional<ConstCFGElementRef> Elem,
+                             const StackFrame *SF, unsigned Count) override {
+    BindElem = Elem;
+    BindSF = SF;
+    BindCount = Count;
+  }
+
+private:
 public:
   RegionStoreManager(ProgramStateManager &mgr)
       : StoreManager(mgr), RBFactory(mgr.getAllocator()),
@@ -583,6 +598,9 @@ public:
 
   LimitedRegionBindingsRef
   removeSubRegionBindings(LimitedRegionBindingsConstRef B, const SubRegion *R);
+  /// PoC(#223400): value to install as the default binding of \p Concrete when
+  /// a bind at a symbolic offset inside it forces us to forget its contents.
+  SVal conjureDefaultForSymbolicOffsetBind(const SubRegion *Concrete);
   std::optional<SVal>
   getConstantValFromConstArrayInitializer(RegionBindingsConstRef B,
                                           const ElementRegion *R);
@@ -1098,6 +1116,18 @@ collectSubRegionBindings(SmallVectorImpl<BindingPair> &Bindings,
                            IncludeAllDefaultBindings);
 }
 
+SVal RegionStoreManager::conjureDefaultForSymbolicOffsetBind(
+    const SubRegion *Concrete) {
+  // Mirrors what InvalidateRegionsWorker::VisitCluster() does: a conjured
+  // symbol as the cluster default makes every later load of a concrete
+  // element produce a *stable* SymbolDerived, instead of UnknownVal.
+  // The type of the symbol is irrelevant; it only serves as a parent symbol.
+  if (!BindElem)
+    return UnknownVal();
+  return svalBuilder.conjureSymbolVal(Concrete, *BindElem, BindSF, Ctx.IntTy,
+                                      BindCount);
+}
+
 LimitedRegionBindingsRef
 RegionStoreManager::removeSubRegionBindings(LimitedRegionBindingsConstRef B,
                                             const SubRegion *Top) {
@@ -1115,7 +1145,8 @@ RegionStoreManager::removeSubRegionBindings(LimitedRegionBindingsConstRef B,
     // sure we don't treat the base region as uninitialized anymore.
     if (TopKey.hasSymbolicOffset()) {
       const SubRegion *Concrete = TopKey.getConcreteOffsetRegion();
-      return B.addBinding(Concrete, BindingKey::Default, UnknownVal());
+      return B.addBinding(Concrete, BindingKey::Default,
+                          conjureDefaultForSymbolicOffsetBind(Concrete));
     }
     return B;
   }
@@ -1135,7 +1166,7 @@ RegionStoreManager::removeSubRegionBindings(LimitedRegionBindingsConstRef B,
   if (TopKey.hasSymbolicOffset()) {
     const SubRegion *Concrete = TopKey.getConcreteOffsetRegion();
     Result = Result.add(BindingKey::Make(Concrete, BindingKey::Default),
-                        UnknownVal());
+                        conjureDefaultForSymbolicOffsetBind(Concrete));
   }
 
   if (Result.isEmpty())
diff --git a/clang/test/Analysis/store-dump-orders.cpp b/clang/test/Analysis/store-dump-orders.cpp
index d99f581f00fe..1a9a789bacc9 100644
--- a/clang/test/Analysis/store-dump-orders.cpp
+++ b/clang/test/Analysis/store-dump-orders.cpp
@@ -47,7 +47,7 @@ void test_output(int n) {
   // CHECK-NEXT:      { "kind": "Direct", "offset": 384, "value": "3 S32b" }
   // CHECK-NEXT:    ]},
   // CHECK-NEXT:    { "cluster": "objsecond", "pointer": "0x{{[0-9a-f]+}}", "items": [
-  // CHECK-NEXT:      { "kind": "Default", "offset": 0, "value": "Unknown" },
+  // CHECK-NEXT:      { "kind": "Default", "offset": 0, "value": "conj_{{[^"]+}}" },
   // CHECK-NEXT:      { "kind": "Direct", "offset": null, "value": "20 S32b" }
   // CHECK-NEXT:    ]}
   // CHECK-NEXT:  ]},
```

**Comment by NagyDonat** [comment]
Thanks for the info!

> The problem is that we don't have everything to conjure symbols from the Store, so those must be smuggled into the context (StoreManager::Bind) via parameters and there are about ~19 call sites to ProgramState::bindLoc that would need to be changed.

Perhaps we should look for a more general solution here :fire:

Each time `SymbolManager::conjureSymbol` is called, its arguments `ConstCFGElementRef Elem`, `const StackFrame *SF` and `unsigned VisitCount` always receive the values returned by `ExprEngine::getCFGElementRef()`, `ExprEngine::getCurrStackFrame()` and `ExprEngine::getNumVisitedCurrent()`. This is not just accidental: we always conjure a symbol within the analysis step where it "appears" and in that analysis step the canonical current `CFGElementRef`, `StackFrame` and `NumVisitedCurrent` can be queried by these methods of the `ExprEngine` [1].

I think we should just store a pointer to the `ExprEngine` singleton among the members of `SymbolManager` and let `conjureSymbol` query these values from the `ExprEngine` whenever it needs them. This would get rid of lots of boilerplate: we wouldn't need to pass around these needed-for-conjuring values through the parameter list of many nested functions.

@steakhal What do you think?

-----

[1] To be precise, I know about two minor discrepancies:
- The current `StackFrame` is often queried from a recent `ExplodedNode` and that can differ from `ExprEngine::getCurrStackFrame()` within the `BeginFunction` callback – but we cannot conjure symbols at that time, because `currStmtIdx` is unset (we are not in the middle of a `CFGBlock`) and therefore we don't have a `ConstCFGElementRef`.
- The loop widening step conjures symbols during the evaluation of a terminator statement of a block (when `currStmtIdx` is unset so we cannot get a "real" a CFG element ref) and uses a hack that constructs a `CFGElementRef` corresponding to the _first statement of the current block_. We could either preserve this hack within `SymbolManager` or perhaps introduce a slightly less ugly solution (e.g. constructing a `ConstCFGElementRef` with `Index == SIZE_MAX` to represent "there is no valid statement index right now").

**Comment by steakhal** [comment]
>I think we should just store a pointer to the ExprEngine singleton among the members of SymbolManager and let conjureSymbol query these values from the ExprEngine whenever it needs them. This would get rid of lots of boilerplate: we wouldn't need to pass around these needed-for-conjuring values through the parameter list of many nested functions.
>
>@steakhal What do you think?

I think it was an intentional choice that SymbolManager doesn't take a reference to ExprEngine. While that choice is founded, it makes the call-sites of conjure cumbersome - as we have seen. I'd rather see a solution for this that minimizes the exposure and unwarranted dependencies on the ExprEngine. My first idea would be to explore how often are these 3 variables (that we use for conjuring and queried from ExprEngine) used together vs separately for other reasons.
For example, if we find that they are always used together, then it would make sense to me to bundle them into a struct (let's call it `ConjureContext` for now) and just pass in a reference to that. That way we don't just pass the glorified god class (ExprEngine) to SymbolManager but just this tiny piece of it - which should be fine.

---

Thanks for these pieces. Really thoughtful.
> [1] To be precise, I know about two minor discrepancies:
>
> - The current StackFrame is often queried from a recent ExplodedNode and that can differ from ExprEngine::getCurrStackFrame() within the BeginFunction callback – but we cannot conjure symbols at that time, because currStmtIdx is unset (we are not in the middle of a CFGBlock) and therefore we don't have a ConstCFGElementRef.

I would have expected that I can conjure inside the `BeginFunction` callback. If this we truly can't meaningfully conjure symbols there, then that's a bug.

> - The loop widening step conjures symbols during the evaluation of a terminator statement of a block (when currStmtIdx is unset so we cannot get a "real" a CFG element ref) and uses a hack that constructs a CFGElementRef corresponding to the first statement of the current block. We could either preserve this hack within SymbolManager or perhaps introduce a slightly less ugly solution (e.g. constructing a ConstCFGElementRef with Index == SIZE_MAX to represent "there is no valid statement index right now").

I'm open to ideas for fixing this. This pitched idea also sounds good. I think differentiating from a usual (valid) CFG element makes sense. Albeit, it would poison all of the call-sites that expect that they can index into the CFGElements array using this index - as those would need to be pessimistically check for bounds from now on - if we go with this approach.
Maybe this "bug" isn't worth fixing after all?

**Comment by NagyDonat** [comment]
> I think it was an intentional choice that SymbolManager doesn't take a reference to ExprEngine.

I agree that it makes some sense to limit access to these huge singletons.

> For example, if we find that they are always used together, then it would make sense to me to bundle them into a struct (let's call it `ConjureContext` for now) and just pass in a reference to that. That way we don't just pass the glorified god class (ExprEngine) to SymbolManager but just this tiny piece of it - which should be fine.

I strongly support this idea :)

There are dozens of `conjure` calls that all use them together. Bundling them together would also make sense from a theoretical POV: these are the "coordinates" that together answer the question "where are we right now?". (This is why they are used for identifying the conjured symbols.)

The ideal pattern would be that they are updated/stored by the engine soon after picking an item from the `WorkList` and starting to work on it; and  then during the handling of that worklist item they are easily available practically everywhere. There are very few places (in the beginning and end of function calls) when it makes sense to update them within the middle of a `WorkList` item and the `ExprEngine` should be able to do so – but other places like the `SymbolManager` only deserve read-only access.

------

> Albeit, it would poison all of the call-sites that expect that they can index into the CFGElements array using this index - as those would need to be pessimistically check for bounds from now on - if we go with this approach.

I checked that the `ConstCFGElementRef` owned by the `SymbolConjured` is only used as an opaque identifier (part of the `Profile`) and by some debug printouts that already check whether it is valid or not. We should just document that it does not always reference a valid statement and then it is completely fine to use CFG element refs with Index == SIZE_MAX to represent the terminator or "between blocks" situations.

**Comment by steakhal** [comment]
I think we are aligned.

---

> Albeit, it would poison all of the call-sites that expect that they can index into the CFGElements array using this index - as those would need to be pessimistically check for bounds from now on - if we go with this approach.
I checked that the ConstCFGElementRef owned by the SymbolConjured is only used as an opaque identifier (part of the Profile) and by some debug printouts that already check whether it is valid or not. We should just document that it does not always reference a valid statement and then it is completely fine to use CFG element refs with Index == SIZE_MAX to represent the terminator or "between blocks" situations.

I also had a look and the use of this index is indeed very limited. It should be possible to change this.
